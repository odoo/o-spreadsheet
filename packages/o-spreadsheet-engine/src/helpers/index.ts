export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
