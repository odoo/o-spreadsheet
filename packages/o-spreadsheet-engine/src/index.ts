export { FunctionCodeBuilder, Scope, type FunctionCode } from "./formulas/code_builder";
export {
  OPERATOR_MAP,
  UNARY_OPERATOR_MAP,
  compile,
  compileTokens,
  functionCache,
} from "./formulas/compiler";
export {
  OP_PRIORITY,
  convertAstNodes,
  iterateAstNodes,
  mapAst,
  parse,
  parseTokens,
} from "./formulas/parser";
export type {
  AST,
  ASTFuncall,
  ASTOperation,
  ASTString,
  ASTSymbol,
  ASTUnaryOperation,
} from "./formulas/parser";
export { rangeTokenize } from "./formulas/range_tokenizer";
export { POSTFIX_UNARY_OPERATORS, tokenize } from "./formulas/tokenizer";
export type { Token } from "./formulas/tokenizer";
export { FunctionRegistry } from "./functions/function_registry";
export { categories } from "./functions/functionRegistry";
export * from "./helpers";
export { Model } from "./model";
export { BasePlugin } from "./plugins/base_plugin";
export { Registry } from "./registry";
export { StateObserver, type StateObserverChange } from "./state_observer";
export * from "./translation";
export { CellValue } from "./types/cells";
export { SpreadsheetClipboardData } from "./types/clipboard";
export { CoreGetters, PluginGetters } from "./types/coreGetters";
export * from "./types/errors";
export { Format } from "./types/format";
export * from "./types/history";
export * from "./types/locale";
export * from "./types/misc";
export * from "./types/validator";
export const __info__ = {};
