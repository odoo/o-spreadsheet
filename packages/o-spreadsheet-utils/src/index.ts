export * from "./translation";
export * from "./uuid";
