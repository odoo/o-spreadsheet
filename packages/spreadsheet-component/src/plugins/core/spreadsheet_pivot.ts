import { ApplyRangeChange, Range } from "../../types";
import { CorePlugin } from "../core_plugin";

function adaptPivotRange(
  range: Range | undefined,
  applyChange: ApplyRangeChange
): Range | undefined {
  if (!range) {
    return undefined;
  }
  const change = applyChange(range);
  switch (change.changeType) {
    case "NONE":
      return range;
    case "REMOVE":
      return undefined;
    default:
      return change.range;
  }
}

export class SpreadsheetPivotCorePlugin extends CorePlugin {
  adaptRanges(applyChange: ApplyRangeChange) {
    for (const pivotId of this.getters.getPivotIds()) {
      const definition = this.getters.getPivotCoreDefinition(pivotId);
      if (definition.type !== "SPREADSHEET") {
        continue;
      }
      if (definition.dataSet) {
        const { sheetId, zone } = definition.dataSet;
        const range = this.getters.getRangeFromZone(sheetId, zone);
        const adaptedRange = adaptPivotRange(range, applyChange);

        const dataSet = adaptedRange && {
          sheetId: adaptedRange.sheetId,
          zone: adaptedRange.zone,
        };
        this.dispatch("UPDATE_PIVOT", { pivotId, pivot: { ...definition, dataSet } });
      }
    }
  }
}
