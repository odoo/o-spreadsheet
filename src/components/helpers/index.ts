export * from "@odoo/o-spreadsheet-engine/components/helpers/css";
