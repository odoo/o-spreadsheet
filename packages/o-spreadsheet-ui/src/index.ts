import * as Engine from "@odoo/o-spreadsheet-engine";

export function salut() {
  console.log("Salut !");
}

export function createSpreadsheetModel() {
  return Engine.SpreadsheetModel();
}