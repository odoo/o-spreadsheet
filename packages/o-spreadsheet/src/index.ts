import * as UI from "@odoo/o-spreadsheet-ui";


export { UI };