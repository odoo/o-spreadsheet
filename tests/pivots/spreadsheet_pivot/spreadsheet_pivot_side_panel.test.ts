import { Model, SpreadsheetChildEnv } from "../../../src";
import { toZone } from "../../../src/helpers";
import { SpreadsheetPivot } from "../../../src/helpers/pivot/spreadsheet_pivot/spreadsheet_pivot";
import { createSheet, setCellContent, undo } from "../../test_helpers/commands_helpers";
import { click, dragElement, setInputValueAndTrigger } from "../../test_helpers/dom_helper";
import { getCellText } from "../../test_helpers/getters_helpers";
import { mountSpreadsheet, nextTick } from "../../test_helpers/helpers";
import { mockGetBoundingClientRect } from "../../test_helpers/mock_helpers";
import { SELECTORS, addPivot, updatePivot } from "../../test_helpers/pivot_helpers";

describe("Spreadsheet pivot side panel", () => {
  let model: Model;
  let fixture: HTMLElement;
  let env: SpreadsheetChildEnv;

  beforeEach(async () => {
    ({ env, model, fixture } = await mountSpreadsheet());
    setCellContent(model, "A1", "Customer");
    setCellContent(model, "B1", "Product");
    setCellContent(model, "C1", "Amount");
    setCellContent(model, "A2", "Alice");
    setCellContent(model, "B2", "Chair");
    setCellContent(model, "C2", "10");
    setCellContent(model, "A3", "Bob");
    setCellContent(model, "B3", "Table");
    setCellContent(model, "C3", "20");
    addPivot(model, "A1:C3", {}, "1");
    env.openSidePanel("PivotSidePanel", { pivotId: "1" });
    await nextTick();
  });

  test("It should correctly be displayed", async () => {
    expect(fixture.querySelector(".o-sidePanel")).toMatchSnapshot();
  });

  test("It should display only the selection input when the dataSet is not valid", async () => {
    updatePivot(model, "1", { dataSet: undefined });
    await nextTick();
    expect(fixture.querySelector(".o-sidePanel")).toMatchSnapshot();
  });

  test("It should be able to change the pivot name", async () => {
    await click(fixture.querySelector(".pivot-defer-update input")!);
    await nextTick();
    await click(fixture.querySelector(".o_sp_en_rename")!);
    const input = fixture.querySelector(".o_sp_en_name") as HTMLInputElement;
    input.value = "New Pivot Name";
    input.dispatchEvent(new Event("input"));
    await click(fixture.querySelector(".o_sp_en_save")!);
    expect(model.getters.getPivotName("1")).toEqual("New Pivot Name");
  });

  test("It should be able to defer updates", async () => {
    setCellContent(model, "A1", "amount");
    setCellContent(model, "A2", "10");
    setCellContent(model, "A3", "20");
    addPivot(model, "A1:A3", {}, "3");
    env.openSidePanel("PivotSidePanel", { pivotId: "3" });
    await nextTick();
    await click(fixture.querySelector(".pivot-defer-update input")!);
    await click(fixture.querySelector(".add-dimension")!);
    expect(fixture.querySelector(".o-popover")).toBeDefined();
    await click(fixture, ".o-autocomplete-value");
    expect(model.getters.getPivotCoreDefinition("3").columns).toEqual([]);
    await click(fixture.querySelector(".sp_apply_update")!);
    expect(model.getters.getPivotCoreDefinition("3").columns).toEqual([
      { name: "amount", order: "asc" },
    ]);
  });

  test("it should not defer update when the dataset is updated", async () => {
    await click(fixture.querySelector(".pivot-defer-update input")!);
    expect((fixture.querySelector(".pivot-defer-update input")! as HTMLInputElement).checked).toBe(
      true
    );
    const input = fixture.querySelector(".o-selection-input input") as HTMLInputElement;
    input.value = "A1:C2";
    input.dispatchEvent(new Event("input"));
    await nextTick();
    await click(fixture.querySelector(".o-selection-ok")!);
    expect(model.getters.getPivotCoreDefinition("1")["dataSet"].zone).toEqual(toZone("A1:C2"));
  });

  test("It should hide range error message if the user updates the range even if the pivot is dirty", async () => {
    updatePivot(model, "1", { dataSet: undefined });
    await nextTick();
    const pivot = model.getters.getPivot("1") as SpreadsheetPivot;
    expect(fixture.querySelector(".sp_range_error_message")?.textContent).toEqual(
      pivot.invalidRangeMessage
    );
    const input = fixture.querySelector(".o-selection-input input") as HTMLInputElement;
    input.value = "A1:C2";
    input.dispatchEvent(new Event("input"));
    await nextTick();
    await click(fixture.querySelector(".o-selection-ok")!);
    expect(fixture.querySelectorAll(".sp_range_error_message")).toHaveLength(0);
  });

  test("Can rename a pivot", async () => {
    await click(fixture.querySelector(".o_sp_en_rename")!);
    const input = fixture.querySelector(".o_sp_en_name") as HTMLInputElement;
    input.value = "New Pivot Name";
    input.dispatchEvent(new Event("input"));
    await click(fixture.querySelector(".o_sp_en_save")!);
    expect(model.getters.getPivotName("1")).toEqual("New Pivot Name");
  });

  test("Can duplicate a pivot and undo the whole action with one step backward", async () => {
    await click(fixture, SELECTORS.COG_WHEEL);
    await click(fixture, SELECTORS.DUPLICATE_PIVOT);
    const pivotId = model.getters.getPivotId("2")!;
    expect(model.getters.getPivot(pivotId)).toBeDefined();
    expect(model.getters.getPivotDisplayName(pivotId)).toEqual("(#2) Pivot (copy)");
    expect(fixture.querySelector(".o-sidePanelTitle")?.textContent).toEqual("Pivot #2");
    expect(model.getters.getSheetIds()).toHaveLength(2);
    expect(model.getters.getSheetName(model.getters.getActiveSheetId())).toBe(
      "Pivot (copy) (Pivot #2)"
    );
    expect(getCellText(model, "A1")).toBe("=PIVOT(2)");

    undo(model);

    expect(model.getters.getPivotId("2")).toBeUndefined();
    expect(model.getters.getSheetIds()).toHaveLength(1);
    expect(model.getters.getSheetName(model.getters.getActiveSheetId())).toBe("Sheet1");
  });

  test("Can duplicate a pivot when a duplicate sheet name already exists", async () => {
    createSheet(model, { name: "Pivot (copy) (Pivot #2)" });
    await click(fixture, SELECTORS.COG_WHEEL);
    await click(fixture, SELECTORS.DUPLICATE_PIVOT);
    const pivotId = model.getters.getPivotId("2")!;
    expect(model.getters.getPivot(pivotId)).toBeDefined();
    expect(model.getters.getPivotDisplayName(pivotId)).toEqual("(#2) Pivot (copy)");
    expect(fixture.querySelector(".o-sidePanelTitle")?.textContent).toEqual("Pivot #2");
    expect(model.getters.getSheetIds()).toHaveLength(3);
    expect(model.getters.getSheetName(model.getters.getActiveSheetId())).toBe(
      "Pivot (copy) (Pivot #2) (1)"
    );
    expect(getCellText(model, "A1")).toBe("=PIVOT(2)");
  });

  test("Can flip axes of a pivot", async () => {
    updatePivot(model, "1", {
      rows: [{ name: "Contact Name", order: "asc" }],
      columns: [{ name: "Active", order: "asc" }],
    });
    await click(fixture, SELECTORS.COG_WHEEL);
    await click(fixture, SELECTORS.FLIP_AXIS_PIVOT);
    const pivotId = model.getters.getPivotId("1")!;
    expect(model.getters.getPivotCoreDefinition(pivotId).rows).toEqual([
      { name: "Active", order: "asc" },
    ]);
    expect(model.getters.getPivotCoreDefinition(pivotId).columns).toEqual([
      { name: "Contact Name", order: "asc" },
    ]);
  });

  test("Pivot dimensions are ordered 'asc' by default", async () => {
    setCellContent(model, "A1", "amount");
    setCellContent(model, "A2", "10");
    setCellContent(model, "A3", "20");
    addPivot(model, "A1:A3", {}, "3");
    env.openSidePanel("PivotSidePanel", { pivotId: "3" });
    await nextTick();
    await click(fixture.querySelector(".add-dimension")!);
    expect(fixture.querySelector(".o-popover")).toBeDefined();
    await click(fixture, ".o-autocomplete-value");
    expect(model.getters.getPivotCoreDefinition("3").columns).toEqual([
      { name: "amount", order: "asc" },
    ]);
  });

  test("should reset side panel if discard is clicked", async () => {
    await click(fixture.querySelector(".pivot-defer-update input")!);
    expect(fixture.querySelectorAll(".pivot-dimension")).toHaveLength(0);
    await click(fixture.querySelector(".add-dimension")!);
    expect(fixture.querySelector(".o-popover")).toBeDefined();
    await click(fixture.querySelectorAll(".o-autocomplete-value")[0]);
    expect(fixture.querySelectorAll(".pivot-dimension")).toHaveLength(1);
    await click(fixture.querySelector(".fa-undo")!);
    expect(fixture.querySelectorAll(".pivot-dimension")).toHaveLength(0);
  });

  test("Measures have the correct default aggregator", async () => {
    setCellContent(model, "A1", "amount");
    setCellContent(model, "A2", "10");
    setCellContent(model, "B1", "person");
    setCellContent(model, "B2", "Alice");
    addPivot(model, "A1:B2", {}, "3");
    env.openSidePanel("PivotSidePanel", { pivotId: "3" });
    await nextTick();

    await click(fixture.querySelector(".o-pivot-measure .add-dimension")!);
    await click(fixture.querySelectorAll(".o-autocomplete-value")[0]);
    expect(model.getters.getPivotCoreDefinition("3").measures).toEqual([
      { name: "amount", aggregator: "sum" },
    ]);

    await click(fixture.querySelector(".o-pivot-measure .add-dimension")!);
    await click(fixture.querySelectorAll(".o-autocomplete-value")[1]);
    expect(model.getters.getPivotCoreDefinition("3").measures).toEqual([
      { name: "amount", aggregator: "sum" },
      { name: "person", aggregator: "count" },
    ]);
  });

  test("should preserve the sorting of the dimension after ordering is changed", async () => {
    mockGetBoundingClientRect({
      /**
       * 'pt-1' is the class of the main div of the pivot dimension
       */
      "pt-1": () => ({
        height: 10,
        y: 0,
      }),
      "o-section-title": () => ({
        height: 10,
        y: 10,
      }),
      "pivot-dimensions": () => ({
        height: 40,
        y: 0,
      }),
    });
    await click(fixture.querySelector(".add-dimension")!);
    await click(fixture.querySelectorAll(".o-autocomplete-value")[0]);
    await setInputValueAndTrigger(fixture.querySelector(".pivot-dimension select"), "desc");
    expect(model.getters.getPivotCoreDefinition("1").columns).toEqual([
      { name: "Amount", order: "desc" },
    ]);
    await dragElement(fixture.querySelector(".pivot-dimension")!, { x: 0, y: 30 }, undefined, true);
    expect(model.getters.getPivotCoreDefinition("1").rows).toEqual([
      { name: "Amount", order: "desc" },
    ]);
  });

  test("Invalid pivot dimensions are displayed as such in the side panel", async () => {
    setCellContent(model, "A1", "ValidDimension");
    setCellContent(model, "A2", "10");
    addPivot(model, "A1:A2", {
      columns: [{ name: "ValidDimension" }],
      rows: [{ name: "InvalidDimension" }],
    });
    env.openSidePanel("PivotSidePanel", { pivotId: "1" });
    await nextTick();
    const pivotDimensionEls = fixture.querySelectorAll<HTMLElement>(".pivot-dimension")!;
    const validDimensionEl = pivotDimensionEls[0];
    expect(validDimensionEl.classList).not.toContain("pivot-dimension-invalid");
    expect(validDimensionEl.querySelector(".fa-exclamation-triangle")).toBe(null);

    const invalidDimensionEl = pivotDimensionEls[1];
    expect(invalidDimensionEl.classList).toContain("pivot-dimension-invalid");
    expect(invalidDimensionEl.querySelector(".fa-exclamation-triangle")).not.toBe(null);
  });
});
