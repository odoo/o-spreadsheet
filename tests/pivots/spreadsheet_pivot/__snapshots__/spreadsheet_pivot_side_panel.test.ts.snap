// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spreadsheet pivot side panel It should correctly be displayed 1`] = `
<div
  class="o-sidePanel"
>
  <div
    class="o-sidePanelHeader"
  >
    <div
      class="o-sidePanelTitle"
    >
      Pivot #1
    </div>
    <div
      class="o-sidePanelClose"
    >
      ✕
    </div>
  </div>
  <div
    class="o-sidePanelBody-container d-flex flex-grow-1 "
  >
    <div
      class="o-sidePanel-handle-container"
    >
      <div
        class="o-sidePanel-handle"
      >
        <svg
          class="o-icon"
          fill="currentColor"
          viewBox="0 0 4 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="2"
            cy="3.5"
            r="1"
          />
          <circle
            cx="2"
            cy="6.5"
            r="1"
          />
          <circle
            cx="2"
            cy="9.5"
            r="1"
          />
          <circle
            cx="2"
            cy="12.5"
            r="1"
          />
        </svg>
      </div>
    </div>
    <div
      class="o-sidePanelBody"
    >
      <div
        class="d-flex flex-column h-100 justify-content-between overflow-hidden"
      >
        <div
          class="h-100 overflow-auto"
        >
          <div
            class="o-section"
          >
            <div
              class="o-section-title"
            >
              <div
                class="d-flex flex-row justify-content-between align-items-center"
              >
                 Name 
                <span
                  class="fa fa-cog os-cog-wheel-menu-icon"
                />
                
                
              </div>
            </div>
            
            
            <div
              class="o_sp_en_display_name"
            >
              (#1) Pivot
            </div>
            <div
              class="btn btn-link o_sp_en_rename"
            >
              <i
                class="fa fa-pencil me-1"
              />
               Rename 
            </div>
            <br />
            
          </div>
          
          <div
            class="o-section"
          >
            <div
              class="o-section-title"
            >
              Range
            </div>
            
            <div
              class="o-selection"
            >
              <div
                class="o-selection-input d-flex flex-row"
              >
                <div
                  class="position-relative w-100"
                >
                  <input
                    class="w-100 o-required"
                    spellcheck="false"
                    style="color: #000;"
                    type="text"
                  />
                  
                </div>
                
              </div>
              
              
              <div
                class="o-selection-input d-flex flex-row"
              >
                
                
                
              </div>
            </div>
            
            
          </div>
          
          <div
            class="o_side_panel_section pivot-dimensions o-section"
          >
            <div
              class="fw-bold py-1 d-flex flex-row justify-content-between align-items-center o-section-title"
            >
               Columns 
              <span
                class="btn btn-sm btn-link add-dimension"
              >
                 Add 
              </span>
              
              
            </div>
            
            
            <div
              class="fw-bold pt-4 pb-1 d-flex flex-row justify-content-between align-items-center o-section-title"
            >
               Rows 
              <span
                class="btn btn-sm btn-link add-dimension"
              >
                 Add 
              </span>
              
              
            </div>
            
            
            <div
              class="fw-bold pt-4 pb-1 d-flex flex-row justify-content-between align-items-center o-section-title o-pivot-measure"
            >
               Measures 
              <span
                class="btn btn-sm btn-link add-dimension"
              >
                 Add 
              </span>
              
              
            </div>
            
            
          </div>
          
        </div>
        <div
          class="o-section align-items-center border-top d-flex flex-row justify-content-between py-1 pivot-defer-update"
        >
          
          <label
            class="o-checkbox"
            title="Changing the pivot definition requires to reload the data. It may take some time."
          >
            <input
              class="align-middle"
              type="checkbox"
            />
            Defer updates
            
          </label>
          
          
        </div>
        
      </div>
    </div>
    
  </div>
</div>
`;

exports[`Spreadsheet pivot side panel It should display only the selection input when the dataSet is not valid 1`] = `
<div
  class="o-sidePanel"
>
  <div
    class="o-sidePanelHeader"
  >
    <div
      class="o-sidePanelTitle"
    >
      Pivot #1
    </div>
    <div
      class="o-sidePanelClose"
    >
      ✕
    </div>
  </div>
  <div
    class="o-sidePanelBody-container d-flex flex-grow-1 "
  >
    <div
      class="o-sidePanel-handle-container"
    >
      <div
        class="o-sidePanel-handle"
      >
        <svg
          class="o-icon"
          fill="currentColor"
          viewBox="0 0 4 16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="2"
            cy="3.5"
            r="1"
          />
          <circle
            cx="2"
            cy="6.5"
            r="1"
          />
          <circle
            cx="2"
            cy="9.5"
            r="1"
          />
          <circle
            cx="2"
            cy="12.5"
            r="1"
          />
        </svg>
      </div>
    </div>
    <div
      class="o-sidePanelBody"
    >
      <div
        class="d-flex flex-column h-100 justify-content-between overflow-hidden"
      >
        <div
          class="h-100 overflow-auto"
        >
          <div
            class="o-section"
          >
            <div
              class="o-section-title"
            >
              <div
                class="d-flex flex-row justify-content-between align-items-center"
              >
                 Name 
                <span
                  class="fa fa-cog os-cog-wheel-menu-icon"
                />
                
                
              </div>
            </div>
            
            
            <div
              class="o_sp_en_display_name"
            >
              (#1) Pivot
            </div>
            <div
              class="btn btn-link o_sp_en_rename"
            >
              <i
                class="fa fa-pencil me-1"
              />
               Rename 
            </div>
            <br />
            
          </div>
          
          <div
            class="o-section"
          >
            <div
              class="o-section-title"
            >
              Range
            </div>
            
            <div
              class="o-selection"
            >
              <div
                class="o-selection-input d-flex flex-row"
              >
                <div
                  class="position-relative w-100"
                >
                  <input
                    class="w-100 o-required o-focused o-invalid border-danger position-relative text-decoration-underline"
                    spellcheck="false"
                    style="color: #000;"
                    type="text"
                  />
                  <span
                    class="error-icon text-danger position-absolute d-flex align-items-center"
                    title="This range is invalid"
                  >
                    <div
                      class="o-icon"
                    >
                      <i
                        class="fa fa-exclamation-circle"
                      />
                    </div>
                  </span>
                  
                </div>
                
              </div>
              
              
              <div
                class="o-selection-input d-flex flex-row"
              >
                
                <button
                  class="o-btn-action bg-transparent fw-bold o-selection-ko"
                >
                   Reset 
                </button>
                
                <button
                  class="o-btn-action bg-transparent fw-bold o-selection-ok"
                >
                   Confirm 
                </button>
                
              </div>
            </div>
            <span
              class="text-danger sp_range_error_message"
            >
              The pivot cannot be created because the dataset is missing.
            </span>
            
          </div>
          
          
        </div>
        <div
          class="o-section align-items-center border-top d-flex flex-row justify-content-between py-1 pivot-defer-update"
        >
          
          <label
            class="o-checkbox"
            title="Changing the pivot definition requires to reload the data. It may take some time."
          >
            <input
              class="align-middle"
              type="checkbox"
            />
            Defer updates
            
          </label>
          
          
        </div>
        
      </div>
    </div>
    
  </div>
</div>
`;
